hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileBlock.java

```diff
@@ -1742,15 +1757,21 @@ protected HFileBlock readBlockDataInternal(FSDataInputStream is, long offset,
         onDiskSizeWithHeader = getOnDiskSizeWithHeader(headerBuf, checksumSupport);
       }
 
-      // The common case is that onDiskSizeWithHeader was produced by a read without checksum
-      // validation, so give it a sanity check before trying to use it.
-      if (!checkOnDiskSizeWithHeader(onDiskSizeWithHeader)) {
+      // The common case is that onDiskSizeWithHeader or headerBuf was produced by a read without
+      // checksum validation, so give it a sanity check before trying to use it.
+      // if checksumType on headerBuf validate fail,the headerBuf may corrupted
+      if (
+        !checkOnDiskSizeWithHeader(onDiskSizeWithHeader)
```

Comment:
Instead of doubling up the checks, why not have two separate if-statements with two different error results? You could move the new `checkCheckSumTypeOnHeaderBuf` up to before `onDiskSizeWithHeader` is populated.